papis/commands/doctor.py

```diff
@@ -0,0 +1,374 @@
+
+        return __fix
+
+    for _f in files:
+        if not os.path.exists(_f):
+            results.append(Error(name=FILES_CHECK_NAME,
+                                 path=folder or "",
+                                 msg=("File '{}' declared but does not exist"
+                                      .format(_f)),
+                                 suggestion_cmd=("papis edit --doc-folder {}"
+                                                 .format(folder)),
+                                 fix_action=_fix(_f),
+                                 payload=_f,
+                                 doc=doc))
+    return results
+
+
+KEYS_EXIST_CHECK_NAME = "keys-exist"
+
+
+def keys_check(doc: papis.document.Document) -> List[Error]:
+    """
+    It checks wether the keys provided in the configuration
+    option ``doctor-keys-check`` exit in the document.
+    """
+    keys = papis.config.getlist("doctor-keys-exist-keys")
+    folder = doc.get_main_folder()
+    results = []  # type: List[Error]
+    for k in keys:
+        if k not in doc or len(doc[k]) == 0:
+            results.append(Error(name=KEYS_EXIST_CHECK_NAME,
+                                 path=folder or "",
+                                 msg=("Key '{}' does not exist"
+                                      .format(k)),
+                                 suggestion_cmd=("papis edit --doc-folder {}"
+                                                 .format(folder)),
+                                 fix_action=lambda: None,
+                                 payload=k,
+                                 doc=doc))
+    return results
+
+
+REFS_CHECK_NAME = "refs"
+
+
+def refs_check(doc: papis.document.Document) -> List[Error]:
+    """
+    It checks that a ref exists and if not it
+    tries to create one according to the ref-format configuration
+    if the user chooses to fix it.
+    """
+    folder = doc.get_main_folder()
+    bad_symbols = re.compile(r"[ ,{}\[\]@#`']")
+
+    def _fix() -> None:
+        ref = papis.bibtex.create_reference(doc, force=True)
+        _db = papis.database.get()
+        logger.info("Setting ref '%s' in '%s'",
+                    ref,
+                    papis.document.describe(doc))
+        doc["ref"] = ref
+        doc.save()
+        _db.update(doc)
+
+    if not doc["ref"] or not str(doc["ref"]).strip():
+        return [Error(name=REFS_CHECK_NAME,
+                      path=folder or "",
+                      msg="Reference missing.",
+                      suggestion_cmd=("papis edit --doc-folder {}"
+                                      .format(folder)),
+                      fix_action=_fix,
+                      payload="",
+                      doc=doc)]
+    m = bad_symbols.findall(str(doc["ref"]))
+    if m:
+        return [Error(name=REFS_CHECK_NAME,
+                      path=folder or "",
+                      msg=("Bad characters ({}) found in reference."
+                           .format(set(list(m)))),
```

Comment:
```suggestion
                           .format(set(m))),
```
? Does that not work?

Reply:
yes this works wonderfully thank you!